dds_web/web/user.py

```diff
@@ -0,0 +1,223 @@
+
+
+@auth_blueprint.route("/register", methods=["POST"])
+def register():
+    """Handles the creation of a new user"""
+    form = dds_web.forms.RegistrationForm()
+
+    # Validate form - validators defined in form class
+    if form.validate_on_submit():
+        # Create new user row by loading form data into schema
+        try:
+            new_user = user_schemas.NewUserSchema().load(form.data)
+
+        except marshmallow.ValidationError as valerr:
+            flask.current_app.logger.info(valerr)
+            raise
+        except (sqlalchemy.exc.SQLAlchemyError, sqlalchemy.exc.IntegrityError) as sqlerr:
+            raise ddserr.DatabaseError from sqlerr
+
+        # Go to two factor authentication setup
+        # TODO: Change this after email is introduced
+        flask_login.login_user(new_user)
+        return flask.redirect(flask.url_for("auth_blueprint.two_factor_setup"))
+
+    # Go to registration form
+    return flask.render_template("user/register.html", form=form)
+
+
+@auth_blueprint.route("/login", methods=["GET", "POST"])
+def login():
+    """Log user in with DDS credentials."""
+
+    # Redirect to index if user is already authenticated
+    if flask_login.current_user.is_authenticated:
+        if flask_login.current_user.has_2fa:
+            return flask.redirect(flask.url_for("auth_blueprint.index"))
+        return flask.redirect(flask.url_for("auth_blueprint.two_factor_setup"))
+
+    # Check if for is filled in and correctly (post)
+    form = forms.LoginForm()
+    if form.validate_on_submit():
+        # Get user from database
+        user = models.User.query.get(form.username.data)
+
+        # Unsuccessful login
+        if not user or not user.verify_password(input_password=form.password.data):
+            flask.flash("Invalid username or password.")
+            return flask.redirect(flask.url_for("auth_blueprint.login"))  # Try login again
+
+        # Correct username and password --> log user in
+        flask_login.login_user(user)
+        flask.flash("Logged in successfully.")
+
+        next = flask.request.args.get("next")
+        # is_safe_url should check if the url is safe for redirects.
+        if not dds_web.utils.is_safe_url(next):
+            return flask.abort(400)
+
+        # Go to home page
+        return flask.redirect(next or flask.url_for("auth_blueprint.index"))
+
+    # Go to login form (get)
+    return flask.render_template("user/login.html", form=form)
+
+
+@auth_blueprint.route("/logout", methods=["POST"])
+def logout():
+    """Logout user."""
+
+    if flask_login.current_user.is_authenticated:
+        flask_login.logout_user()
+
+    return flask.redirect(flask.url_for("auth_blueprint.index"))
+
+
+@auth_blueprint.route("/twofactor", methods=["GET"])
+def two_factor_setup():
+    """Setup two factor authentication."""
+    # since this page contains the sensitive qrcode, make sure the browser
+    # does not cache it
+    flask_login.current_user.set_2fa_seen()
```

Comment:
This is never checked at the moment right? And also, I'm wondering how strict we should be. This metod is named 'set_2fa_seen'  but it sets 'has_2fa'. I'm thinking there might be a difference between seeing the token and actually having it set up. We can perhaps discuss this more in a meeting.

Reply:
So the idea is that the secret should only be seen once, because otherwise we can't guarantee that it's the correct user that's actually using the 2fa authentication later on. So it should be seen once, and then never again. If they would need to redo this for some reason there should be another new secret instead, but how this would work and if we need some checks or verification etc, I'm not sure. 

Regarding the `set_2fa_seen` --> not entirely sure why I decided to it like that, probably thought it looked better than `flask_login.current_user.has_2fa = True` and it gives more power to the db to handle it.

Reply:
What do you mean with checked?

Reply:
Ok maybe I'll remove this for now. I anyway have it in two places ish so I think I just missed something somewhere. Will remove this part and it will be (for now) that they can only set it up once.

Reply:
with 'checked' I meant that you set it True but we should perhaps check when the request is made whether it's already set or not?